+bids/+util/update_struct.m

```diff
@@ -0,0 +1,91 @@
+function js = update_struct(js, varargin)
+  %
+  % Updates structure with new values.
+  % Can add new fields, replace field values, remove fields,
+  %  and append  new values to a cellarray.
+  %
+  % Designed for manipulating json structures and will not work
+  % on structarrays.
+  %
+  % USAGE::
+  %
+  %  js = update_struct(key1, value1, key2, value2);
+  %  js = update_struct(struct(key1, value1, ...
+  %                            key2, value2));
+  %
+  % Examples:
+  % ---------
+  % Adding and replacing existing fields:
+  %   update_struct(struct('a', 'val_a'),...
+  %                 'a', 'new_val', 'b', 'val_b')
+  %   struct with fields:
+  %      a: 'new_val'
+  %      b: 'val_b'
+  % Removing field from structure:
+  %   update_struct(struct('a', 'val_a', 'b', 'val_b'),
+  %                 'b', [])
+  %   struct with fields:
+  %      a: 'val_a'
+  % Appending values to existing field:
+  %   update_struct(struct('a', 'val_a', 'b', 'val_b'),
+  %                 'b-add', 'val_b2')
```

Comment:
```suggestion
  %                 'b-add', 'val_b2')
```
Random thought.
I know this is an edge case in the world of bids because we would expect the field to be camel camse but what if one of the field to update is named "b-add", do you have then do something like: "b-add-add" ?

Reply:
maybe it would be better to have simpler function:
- one for adding
- one for removing
- one for appending an element to a an array or a cell
- one for popping an element to an array or a cell

or have the first argument specify what "action" to perform.

Reply:
Well, no matter how BIDS will decide to make life needlessly complicated, but if json is transformed into matlab structure, the fieldnames like `b-add` will be forbidden.

Splitting into several subcommand: `metadata_add`, `metadata-append`, `metadata-rm` can be implemented, and maybe is a good idea. But then it would be difficult to work with struct instead of parameters.

For poping (remove and return) is more complicated, as matlab do not modify current object, and multiple returns looks ugly:
```matlab
 [file, val] = file.metadata_pop('val_a');
```

What is your take on not forcing to read metadata when creating File instance, and making it on request only?

Reply:
> the fieldnames like `b-add` will be forbidden

Oh dear... Matlab 101. Silly me for forgetting this... I guess that's what happens when you start switching back and forth between languages

Reply:
I agree that inheritance should be part of bids (even if I don't like it due to all the complications). My question should the metadata be loaded implicitly or explicitly (`f.set_metadata();`). 

Implicit reading of metadata imposes that the logic of retrieving json files, and reading them happens every time when a new instance is created, even if it's not needed. 

The way how inheritance principle is implemented is an another issue -- it requires the knowledge of the layout of dataset, which is stored in layout structure, but not in File. For me (very personal opinion), inheritance principle should be applied during creation of layout, by setting File.metadata_files, but not in File.

Reply:
Just realized that `set_metadata` relies on `bids.internal.get_meta_list` that tries to implement the inheritance principle (same one use for bids.layout).

```
    function obj = set_metadata(obj)
      if isempty(obj.metadata_files)
        pattern = '^.*%s\\.json$';
        obj.metadata_files = bids.internal.get_meta_list(obj.path, pattern);
      end
      obj.metadata =  bids.internal.get_metadata(obj.metadata_files);
    end
```

I would say that changing this is beyond the scope of this PR which is on updating metadata. Just thought it was good to mention it.

Reply:
> For poping (remove and return) is more complicated, as matlab do not modify current object, and multiple returns looks ugly:
> 
> ```matlab
>  [file, val] = file.metadata_pop('val_a');
> ```

I think I would just return the file object.

```
file = file.metadata_pop('FieldToPopFrom', 'value_to_pop');
```

Reply:
FWI: will open a small PR to add a parameter to `bids.internal.get_meta_list` to decide to use the inheritance principle or not to list metadata files.

It could then be reused here to let user decide what they want to do.